test/shared/kaocha/integration_helpers.clj

```diff
@@ -126,6 +129,123 @@
+  using (read-line) and (println).
+
+  m is the output from `test-dir-setup`
+  args is a collection of arguments after ./bin/kaocha.
+  eg., [\"--watch\" \"--focus\" \"foo\"]
+  
+  Kills the process after 30 seconds if it has not already
+  been exited. Ensures the process is killed immediately after
+  body is executed. Returns the exit code.
+  
+  Executes body in the following environment:
+  *process* is the running Process executing `./bin/kaocha ~@args`
+  *in* is bound to the output stream of this Process
+  - eg., use (read-line) to read the output of ./bin/kaocha
+  *out* is bound to the input stream of this Process
+  - eg., use (println) to send input to ./bin/kaocha
+  
+  Tips:
+  - avoid using `testing` or `is` in ways that can swallow exceptions
+  in body. Try and make the first exception that's thrown contain
+  all the debugging information you need using the helpers in the
+  rest of this file."
+  [m args & body]
+  `(interactive-process* ~m ~args #(do ~@body)))
+
+(comment
+  (zero?
+    (interactive-process {:dir "." :runner (io/file "echo")} ["hello"]
+                         (let [l1 (read-line)
+                               _ (assert (= "hello" l1) (pr-str l1))])))
+  )
+
+(defn read-string-line
+  "Read a line from the current integration test and throw if the process has died."
+  []
+  (or (read-line)
+      (throw (ex-info "Process killed!!" {}))))
+
+(defn expect-lines
+  "Assert that lines, a vector of strings, matches the next lines from the integration process.
+
+  If not, slurps the rest of the output for debugging purposes and throws an exception."
+  [lines]
+  (mapv (fn [l]
+          (let [s (read-string-line)]
+            (or (is (= l s))
+                (throw (ex-info (format "Failed to match %s\nEntire expected: %s\nRest of stream:\n%s"
+                                        (pr-str l) lines (str/split-lines (slurp *in*)))
+                                {})))))
+        lines))
+
+(defn next-line-matches
+  "Checks that the next line from the integration process matches function f.
+  
+  If not, slurps the rest of the output for debugging purposes and throws an exception."
+  [f]
+  (let [s (try (read-string-line)
+               (catch Throwable e
+                 (is nil "Failed next-line-matches call: process ended!!")
+                 (throw e)))]
+    (or (f s)
+        (throw (ex-info (format "Failed next-line-matches call\nFound: %s\nRest of stream:\n%s"
+                                (pr-str s) (str/split-lines (slurp *in*)))
+                        {})))))
+
+(defn read-until
+  "Keep reading from integration process output until f is true.
+  
+  If never true, reports all the strings that failed."
+  [f]
+  ;; can't recur across try, use atom
+  (let [seen (atom [])]
+    (try
+      (loop []
+        (let [s (read-string-line)]
+          (when-not (f s)
+            (swap! seen conj s)
+            (recur))))
+      (catch Throwable e
```

Comment:
If you're just just expecting `ex-info` `Throwable`s, maybe put `clojure.lang.ExceptionInfo`? We also use slingshot, but I can't think of an idiomatic way to make this catch more specific.